packages/plugin-relay/src/schema-builder.ts

```diff
@@ -287,16 +288,17 @@ schemaBuilderProto.node = function node(param, { interfaces, extensions, id, ...
     nodeName = nodeConfig.name;
 
     this.objectField(ref, this.options.relayOptions.idFieldName ?? 'id', (t) =>
-      t.globalID<{}, false, Promise<GlobalIDShape<SchemaTypes>>>({
+      t.globalID<{}, false, MaybePromise<GlobalIDShape<SchemaTypes>>>({
         nullable: false,
         ...this.options.relayOptions.idFieldOptions,
         ...id,
         args: {},
-        resolve: async (parent, args, context, info) => ({
-          type: nodeConfig.name,
+        resolve: (parent, args, context, info) =>
           // eslint-disable-next-line @typescript-eslint/no-unsafe-argument
-          id: await id.resolve(parent, args, context, info),
-        }),
+          this.completeValue(id.resolve(parent, args, context, info), (globalId) => ({
```

Comment:
Love this change.  I had done this for a few performance critical places in the past and was really surprised by how much of an improvement this type of change makes!